src/cdk-experimental/testing/component-harness.ts

```diff
@@ -0,0 +1,220 @@
+ * the internal behavior is not guaranteed and user should not rely on it. Note
+ * that in most cases, you don't need to care about whether an element is
+ * present when loading the element and don't need to set this parameter unless
+ * you do want to check whether the element is present when calling the
+ * searching function. e.g. you want to make sure some element is not there when
+ * loading the element in order to check whether a "ngif" works well.
+ *
+ * @param global Optional. If global is set to true, the selector will match any
+ *     element on the page and is not limited to the root of the harness. If
+ * global is unset or set to false, the selector will only find elements under
+ * the current root.
+ */
+export interface Options {
+  allowNull?: boolean;
+  global?: boolean;
+}
+
+/**
+ * Type narrowing of Options to allow the overloads of ComponentHarness.find to
+ * return null only if allowNull is set to true.
+ */
+interface OptionsWithAllowNullSet extends Options {
+  allowNull: true;
+}
+
+/**
+ * Locator interface
+ */
+export interface Locator {
+  /**
+   * Get the host element of locator.
+   */
+  host(): TestElement;
+
+  /**
+   * Search the first matched test element.
+   * @param css Selector of the test elements.
+   * @param options Optional, extra searching options
+   */
+  find(css: string, options?: Options): Promise<TestElement|null>;
+
+  /**
+   * Search all matched test elements under current root by css selector.
+   * @param css Selector of the test elements.
+   */
+  findAll(css: string): Promise<TestElement[]>;
+
+  /**
+   * Load the first matched Component Harness.
+   * @param componentHarness Type of user customized harness.
+   * @param root Css root selector of the new component harness.
+   * @param options Optional, extra searching options
+   */
+  load<T extends ComponentHarness>(
+    componentHarness: ComponentHarnessType<T>, root: string,
+    options?: Options): Promise<T|null>;
+
+  /**
+   * Load all Component Harnesses under current root.
+   * @param componentHarness Type of user customized harness.
+   * @param root Css root selector of the new component harnesses.
+   */
+  loadAll<T extends ComponentHarness>(
+    componentHarness: ComponentHarnessType<T>, root: string): Promise<T[]>;
+}
+
+/**
+ * Base Component Harness
+ * This base component harness provides the basic ability to locate element and
+ * sub-component harness. It should be inherited when defining user's own
+ * harness.
+ */
+export class ComponentHarness {
+  constructor(private readonly locator: Locator) {}
+
+  /**
+   * Get the host element of component harness.
+   */
+  host(): TestElement {
+    return this.locator.host();
+  }
+
+  /**
+   * Generate a function to find the first matched test element by css
+   * selector.
+   * @param css Css selector of the test element.
+   */
+  protected find(css: string): () => Promise<TestElement>;
```

Comment:
This ensures that we are always rechecking for the element, in case it has been removed from the DOM. e.g.:

```ts
export class MainComponentHarness extends ComponentHarness {
  readonly title = this.find('h1');
  ...
}

it('whatever', async () => {
  const main = load(MainComponentHarness, ...);
  expect(await main.title()).not.toBeNull();
}); 
```